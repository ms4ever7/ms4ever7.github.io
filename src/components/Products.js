import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchProducts, setPrimaryItem } from '../redux/actions/products';

import NoResultsFound from './NoResultsFound';
import Product from './Product';

class Products extends Component {
  static propTypes = {
    products: PropTypes.array,
    selectedCategories: PropTypes.array,
    selectedColors: PropTypes.array,
    fetchProducts: PropTypes.func,
    setPrimaryItem: PropTypes.func
  }

  static defaultProps = {
    products: [],
    selectedCategories: [],
    selectedColors: [],
    fetchProducts: () => {}
  }

  componentDidMount() {
    this.props.fetchProducts();
  }

  get appropriateItems() {
    const { products, selectedCategories, selectedColors } = this.props;

    return products.filter((product) => {
      const { category, colors } = product;

      if ((selectedCategories.length && !selectedCategories.includes(category))
        || (selectedColors.length && !selectedColors.every(color => colors.includes(color)))) {
        return;
      }

      return product;
    });
  }


  get items() {
    const { selectedCategories, selectedColors, setPrimaryItem } = this.props;

    if (!this.appropriateItems.length) {
      return <NoResultsFound selectedCategories={selectedCategories} selectedColors={selectedColors} />;
    }

    return this.appropriateItems.map(item => <Product key={item.id} data={item} setPrimaryItem={setPrimaryItem} className='productItem' />);
  }

  render() {
    const { products } = this.props;

    if (!products || !products.length) {
      return null;
    }

    return (
      <div className='products__wrapper layout layout-align-center-start flex'>
        {this.items}
      </div>
    );
  }
}

const mapStateToProps = ({ products, filters }) => ({
  products: products.items,
  selectedCategories: filters.selectedCategories || [],
  selectedColors: filters.selectedColors || []
});

const mapDispatchToProps = dispatch => ({
  fetchProducts: () => dispatch(fetchProducts()),
  setPrimaryItem: (item, id) => dispatch(setPrimaryItem(item, id))
});

export default connect(mapStateToProps, mapDispatchToProps)(Products);
