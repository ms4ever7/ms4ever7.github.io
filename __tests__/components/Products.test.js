import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Products from '../../src/components/Products';
// import Product from '../../src/components/Product';

describe('Products', () => {
  let wrapper;
  let mockStore;

  beforeEach(() => {
    mockStore = configureMockStore([]);
  });

  it('should render correctly', () => {
    const props = fulfillProps();
    wrapper = shallow(
      <Products />,
      {
        context: { store: mockStore(props) }
      }
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('is null when no data has been passed', () => {
    const items = [{
      id: 'id',
      price: '$67.94',
      category: 'Athletic',
      name: 'Yeezy Boost',
      owner: 'Adidas',
      colors: ['black', 'red', 'grey'],
      primaryImageUrl: 'test',
      primaryColor: 'Red',
      similarProducts: [
        {
          price: '$67.94',
          imageUrl: 'test',
          color: 'Black'
        }
      ]
    }];
    const wrapper = shallow(<Products />, {
      context: { store: mockStore(fulfillProps({ items })) }
    });

    const render = wrapper.dive();

    expect(render.find('.productItem').exists()).toBeTruthy();
  });
});


const fulfillProps = (props = {}) => ({
  products: {
    items: [],
    ...props
  },
  filters: {
    selectCategory: [],
    selectColors: []
  }
});
