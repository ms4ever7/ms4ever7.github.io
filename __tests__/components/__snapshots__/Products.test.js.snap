// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products should render correctly 1`] = `
<Products
  fetchProducts={[Function]}
  products={Array []}
  selectedCategories={Array []}
  selectedColors={Array []}
  setPrimaryItem={[Function]}
/>
`;
